src/ocaml/parser/syntax.rs

```diff
@@ -0,0 +1,482 @@
+                LambdaSyntax::Record(p, a, b.into_iter().map(|s| set_pos(&s, p)).collect())
+            }
+            LambdaSyntax::Setglobal(_, a, b) => {
+                LambdaSyntax::Setglobal(p, set_pos(&a, p).into(), set_pos(&b, p).into())
+            }
+            LambdaSyntax::Seq(_, xs) => {
+                LambdaSyntax::Seq(p, xs.into_iter().map(|s| set_pos(&s, p)).collect())
+            }
+            LambdaSyntax::Makeblock(_, a, b) => {
+                LambdaSyntax::Makeblock(p, a, b.into_iter().map(|s| set_pos(&s, p)).collect())
+            }
+            LambdaSyntax::Let(_, binds, body) | LambdaSyntax::Letrec(_, binds, body) => {
+                let binds = binds
+                    .into_iter()
+                    .map(|(a, b)| (set_pos(&a, p), set_pos(&b, p)))
+                    .collect();
+                let body = set_pos(&body, p).into();
+                match *l {
+                    LambdaSyntax::Let(_, _, _) => LambdaSyntax::Let(p, binds, body),
+                    LambdaSyntax::Letrec(_, _, _) => LambdaSyntax::Letrec(p, binds, body),
+                    _ => unreachable!(),
+                }
+            }
+            LambdaSyntax::Function(_, args, body) => LambdaSyntax::Function(
+                p,
+                args.into_iter().map(|a| set_pos(&a, p)).collect(),
+                set_pos(&body, p).into(),
+            ),
+            LambdaSyntax::Apply(_, func, args) => LambdaSyntax::Apply(
+                p,
+                set_pos(&func, p).into(),
+                args.into_iter().map(|a| set_pos(&a, p)).collect(),
+            ),
+            LambdaSyntax::FallbackPrimitive(_, a, b) => LambdaSyntax::FallbackPrimitive(
+                p,
+                a,
+                b.into_iter().map(|s| set_pos(&s, p)).collect(),
+            ),
+            LambdaSyntax::FallbackLiteral(_, a) => LambdaSyntax::FallbackLiteral(p, a),
+        }
+    }
+
+    fn test_parser(input: &str, expected: Option<LambdaSyntax>) {
+        let span = Span::new(input);
+        let (rest, res) = parse_syntax(span).unwrap();
+        assert!(rest.is_empty());
+        if let Some(expected) = expected {
+            let pos = expected.get_pos();
+            assert_eq!(set_pos(&res, *pos), expected);
+        }
+    }
+
+    macro_rules! test {
+        ($test_func:ident, $input_str:expr) => {
+            #[test]
+            fn $test_func() {
+                test_parser($input_str, None)
+            }
+        };
+        ($test_func:ident, $input_str:expr, $expected_cloj:expr) => {
+            #[test]
+            fn $test_func() {
+                let placeholder_span = Span::new("");
+                let p = Pos::from_upto(placeholder_span, placeholder_span);
+                let expected = ($expected_cloj)(p);
+                test_parser($input_str, Some(expected))
+            }
+        };
+    }
+
+    test!(test_mastermind, include_str!("tests/mastermind.ir"));
```

Comment:
It's probably worth mentioning the version of the compiler used to generate this IR

Reply:
I've also added the original ocaml source file in the test folder for reference, though right now it's not being used as a test yet